jena-arq/src/main/java/org/apache/jena/http/auth/AuthLib.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.jena.http.auth;
+
+import java.net.Authenticator;
+import java.net.PasswordAuthentication;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.net.http.HttpResponse.BodyHandler;
+import java.util.List;
+
+import org.apache.jena.atlas.web.HttpException;
+import org.apache.jena.http.HttpLib;
+import org.apache.jena.riot.web.HttpNames;
+import org.apache.jena.web.HttpSC;
+
+public class AuthLib {
```

Comment:
Shiro does not support digest and they have closed the ticket for it. In the enterprise it's unlikely to be used. It's not great anyway - basic+https is better - but of course has an overhead on users to set up and digest is still so

The digest code already existed so adding it wasn't too hard. Jetty supports it.



The thing about HTTP Basic is that either it is challenge-response, in which case the request must be repeatable or the password must be provided on first call which exposes it if sent to the wrong place.

Digest always requires a challenge-response cycle . That alone makes it unsuitable for code, rather than people.

Shiro is a whole framework and seems to be to be more server-side-centric. It is used in Fuseki-UI but generates lots of build warnings which may go away eventually (v2) but have been there are several years now. Not sure where the project is going.